pages/api/recent-posts.ts

```diff
@@ -0,0 +1,110 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+
+import { ExtendedRecordMap } from 'notion-types'
+import {
+  getBlockParentPage,
+  getBlockTitle,
+  getPageProperty,
+  idToUuid
+} from 'notion-utils'
+
+import * as config from '@/lib/config'
+import { getSiteMap } from '@/lib/get-site-map'
+import { getSocialImageUrl } from '@/lib/get-social-image-url'
+import { getCanonicalPageUrl } from '@/lib/map-page-url'
+
+type BlogPost = {
+  title: string
+  description: string
+  published: number // unix timestamp
+  url: string
+  authors: string[]
+  coverImage: {
+    url: string
+    width: number
+    height: number
+  }
+}
+
+type Response = {
+  posts: BlogPost[]
+}
+
+/**
+ * This handler exposes recent blog posts as a JSON object for displaying post
+ * previews on the rest of the site.
+ */
+export default async function (req: NextApiRequest, res: NextApiResponse) {
+  const siteMap = await getSiteMap()
+  const ttl = 24 * 60 * 60 // 24 hours
+  const data: Response = {
+    posts: []
+  }
+
+  const limit = 5
+
+  for (const pagePath of Object.keys(siteMap.canonicalPageMap)) {
+    const pageId = siteMap.canonicalPageMap[pagePath]
+    const recordMap = siteMap.pageMap[pageId] as ExtendedRecordMap
+    if (!recordMap) continue
+
+    const keys = Object.keys(recordMap?.block || {})
+    const block = recordMap?.block?.[keys[0]]?.value
+    if (!block) continue
+
+    const parentPage = getBlockParentPage(block, recordMap)
+    const isBlogPost =
+      block.type === 'page' &&
+      block.parent_table === 'collection' &&
+      parentPage?.id === idToUuid(config.rootNotionPageId)
+    if (!isBlogPost) {
+      continue
+    }
+
+    const isPublished = getPageProperty<number>('Published', block, recordMap)
+    if (!isPublished) {
+      continue
+    }
+
+    const title = getBlockTitle(block, recordMap) || config.name
+    const description =
```

Comment:
I noticed pretty much every blog post is missing the description, so they all fall back to the site-wide description, which probably isn't great for SEO.

I started to explore extracting an excerpt, but Notion's block-based approach makes it kind of tricky. I'd recommend including a short excerpt in this Description field in Notion. I'm happy to provide backfill data for all of these if it's helpful. At the very least, we should do this for the most recent posts which appear on the site.

For example:

- **What Can Voltage Structures…**: Knowing the structure of the grid—how lines interconnect and what phases loads are on—is vital for efficient grid maintenance and operations, informing applications ranging from fault localization to phase balancing. 
- **Estimating SAIDI…**: How can we estimate SAIDI (System Average Interruption Duration Index)—the average power outage duration experienced across all customers served—from PowerWatch sensor measurements of only a subset of customers?
- **A Clustering Algorithm**: nLine installs power sensors at outlets in homes, small businesses, and social infrastructure. How do we estimate the extent of a grid outage from individual sensor reports?

Reply:
Looking at this right now, I've wanted to add descriptions for these posts from Notion's side.